src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/ConditionalDeleteResourceHandler.cs

```diff
@@ -0,0 +1,114 @@
+using Microsoft.Health.Fhir.Core.Extensions;
+using Microsoft.Health.Fhir.Core.Features.Conformance;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Features.Search;
+using Microsoft.Health.Fhir.Core.Features.Security;
+using Microsoft.Health.Fhir.Core.Messages.Delete;
+
+namespace Microsoft.Health.Fhir.Core.Features.Resources.Delete
+{
+    public class ConditionalDeleteResourceHandler : BaseResourceHandler, IRequestHandler<ConditionalDeleteResourceRequest, DeleteResourceResponse>
+    {
+        private readonly ISearchService _searchService;
+        private readonly IMediator _mediator;
+        private readonly int _conditionalDeleteMaxItems;
+
+        public ConditionalDeleteResourceHandler(
+            IFhirDataStore fhirDataStore,
+            Lazy<IConformanceProvider> conformanceProvider,
+            IResourceWrapperFactory resourceWrapperFactory,
+            ISearchService searchService,
+            IMediator mediator,
+            ResourceIdProvider resourceIdProvider,
+            IAuthorizationService<DataActions> authorizationService,
+            IOptions<CoreFeatureConfiguration> featureConfiguration)
+            : base(fhirDataStore, conformanceProvider, resourceWrapperFactory, resourceIdProvider, authorizationService)
+        {
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+            EnsureArg.IsNotNull(searchService, nameof(searchService));
+            EnsureArg.IsNotNull(featureConfiguration?.Value, nameof(featureConfiguration));
+
+            _searchService = searchService;
+            _mediator = mediator;
+            _conditionalDeleteMaxItems = featureConfiguration.Value.ConditionalDeleteMaxItems;
+        }
+
+        public async Task<DeleteResourceResponse> Handle(ConditionalDeleteResourceRequest request, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(request, nameof(request));
+
+            DataActions dataActions = (request.HardDelete ? DataActions.Delete | DataActions.HardDelete : DataActions.Delete) | DataActions.Read;
+
+            if (await AuthorizationService.CheckAccess(dataActions, cancellationToken) != dataActions)
+            {
+                throw new UnauthorizedFhirActionException();
+            }
+
+            (IReadOnlyCollection<SearchResultEntry> matchedResults, string ct) = await _searchService.ConditionalSearchAsync(request.ResourceType, request.ConditionalParameters, 2, cancellationToken);
+
+            int count = matchedResults.Count;
+            if (count == 0)
+            {
+                return null;
+            }
+            else if (count == 1)
+            {
+                return await _mediator.Send(new DeleteResourceRequest(request.ResourceType, matchedResults.First().Resource.ResourceId, request.HardDelete), cancellationToken);
+            }
+            else
+            {
+                if (!request.DeleteMultiple || _conditionalDeleteMaxItems <= 0)
```

Comment:
Any chance I can convince you to rewrite this if logic?

I think it would be a bit cleaner if first condition we checked is DeleteMultiple on request, and have two different branches with two different calls to ConditionalSearchAsync methods.

first branch would be copy of delete,
second branch would have while loop.

It just strange to fetch 2 results and then fetch (10) default amount of count results in while loop.

Reply:
Will give it a shot. The loop actually does `_conditionalDeleteMaxItems - itemsDeleted` items on the second pass, so its aiming to fetch the remaining 98 items (if the max is 100) then limit to 10 delete requests in parallel.